lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartManager.java

```diff
@@ -67,9 +71,66 @@ private YarnDriverRestartManager(@Parameter(YarnEvaluatorPreserver.class)
     this.previousContainers = null;
   }
 
+  /**
+   * Determines whether the application master has been restarted based on the container ID environment
+   * variable provided by YARN. If that fails, determine whether the application master is a restart
+   * based on the number of previous containers reported by YARN.
+   * @return true if the application master is a restarted instance, false otherwise.
+   */
   @Override
   public boolean isRestart() {
-    // TODO [REEF-508]: Determine restart based on environment variable as provided by YARN.
+    String containerIdString;
+
+    try {
+      final Map<String, String> envs = System.getenv();
+      containerIdString = envs.get(ApplicationConstants.Environment.CONTAINER_ID.key());
+    } catch (Exception e) {
+      LOG.log(Level.WARNING, "Unable to get the container ID from the environment, exception " +
+          e + " was thrown.");
+      containerIdString = null;
+    }
+
+    if (containerIdString == null) {
+      // container id should always be set in the env by the framework
```

Comment:
We can always recover by using the alternative method, I don't think we should throw here.